packages/opentelemetry-plugin-xml-http-request/src/xhr.ts

```diff
@@ -0,0 +1,463 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  BasePlugin,
+  isUrlIgnored,
+  isWrapped,
+  otperformance,
+  urlMatches,
+} from '@opentelemetry/core';
+import * as tracing from '@opentelemetry/tracing';
+import * as types from '@opentelemetry/types';
+import {
+  addSpanNetworkEvent,
+  getResource,
+  parseUrl,
+  PerformanceTimingNames as PTN,
+} from '@opentelemetry/web';
+import * as shimmer from 'shimmer';
+import { AttributeNames } from './enums/AttributeNames';
+import { EventNames } from './enums/EventNames';
+import { Format } from './enums/Format';
+import {
+  OpenFunction,
+  PropagateTraceHeaderCorsUrls,
+  SendBody,
+  SendFunction,
+  XhrMem,
+} from './types';
+
+/**
+ * XMLHttpRequest config
+ */
+export interface XMLHttpRequestPluginConfig extends types.PluginConfig {
+  // urls which should include trace headers when origin doesn't match
+  propagateTraceHeaderCorsUrls?: PropagateTraceHeaderCorsUrls;
+}
+
+/**
+ * This class represents a XMLHttpRequest plugin for auto instrumentation
+ */
+export class XMLHttpRequestPlugin extends BasePlugin<XMLHttpRequest> {
+  readonly component: string = 'xml-http-request';
+  // @TODO align this with all packages #600
+  readonly version: string = '0.3.0';
+  moduleName = this.component;
+
+  protected _config!: XMLHttpRequestPluginConfig;
+
+  private _tasksCount = 0;
+  private _xhrMem = new WeakMap<XMLHttpRequest, XhrMem>();
+  private _usedResources: { [key: string]: PerformanceResourceTiming[] } = {};
+
+  constructor(config: XMLHttpRequestPluginConfig = {}) {
+    super();
+    this._config = config;
+  }
+
+  /**
+   * Adds custom headers to XMLHttpRequest
+   * @param xhr
+   * @param span
+   * @private
+   */
+  private _addHeaders(xhr: XMLHttpRequest, span: types.Span) {
+    let propagateTraceHeaderUrls =
+      this._config.propagateTraceHeaderCorsUrls || [];
+    if (
+      typeof propagateTraceHeaderUrls === 'string' ||
+      propagateTraceHeaderUrls instanceof RegExp
+    ) {
+      propagateTraceHeaderUrls = [propagateTraceHeaderUrls];
+    }
+
+    const url = (span as tracing.Span).name;
+    const spanUrl = parseUrl(url);
+    let addHeaderOnDifferentOrigin = false;
+
+    if (spanUrl.origin !== window.location.origin) {
+      for (const propagateTraceHeaderUrl of propagateTraceHeaderUrls) {
+        if (urlMatches(url, propagateTraceHeaderUrl)) {
+          addHeaderOnDifferentOrigin = true;
+          break;
```

Comment:
Optional: what would you think about a function say called `_shouldPropagateTraceHeaders` that would return true/false and could be used in an if statement about whether to return early or propagate the headers? Could be slightly cleaner than incorporating the logic into this function.